### DISCUSSION TITLE
MQTT non funziona dopo aggiornamento

### DISCUSSION THREAD

**Discussion by fdlab01** [indomus/forum#670] [GENERALE - Discussioni generiche]
Salve a tutti, da più di un anno utilizzo una configurazione tra Home Assistant Core ed  MQTT Broker Mosquitto su Raspberry, ho anche tre ESP8266 e due shelly collegati al MQTT Broker su raspberry senza problemi.
Dopo l'aggiornamento tutti i dispositivi MQTT non funzionano più, ho seguito diversi issue tracker online per risolvere il problema ma non ne sono venuto a capo... 
Pensando si trattasse di un problema di MQTT o di incasinamento di Raspbian ho effettuato anche un installazione pulita di Raspbian su una nuova micro sd ed ho installato MQTT pulito e configurato a dovere con username e password file ma niente.. se da terminale digito "mosquitto -v" mi restituisce le seguenti linee 

pi@raspberrypi:~ $ mosquitto -v
1630502750: mosquitto version 2.0.11 starting
1630502750: Using default config.
1630502750: Starting in local only mode. Connections will only be possible from clients running on this machine.
1630502750: Create a configuration file which defines a listener to allow remote access.
1630502750: For more details see https://mosquitto.org/documentation/authentication-methods/
1630502750: Opening ipv4 listen socket on port 1883.
1630502750: Error: Address already in use
1630502750: Opening ipv6 listen socket on port 1883.
1630502750: mosquitto version 2.0.11 running

il fatto è che io il file mosquitto.conf l'ho creato e configurato con metodo di autentificazione username password
Qualcuno sa darmi una mano?

**Comment by KC-inDomus** [comment]
1630502750: Opening ipv4 listen socket on port 1883.
1630502750: Error: **Address already in use**

**Reply by fdlab01** [reply]
Grazie per la risposta.. Pensi che il problema sia sul router?? Ho un ip statico assegnato al raspberry e la porta 1883 aperta sia in up che down.
Comunque a prescindere da questo perchè continua a dirmi:

1630502750: Starting in local only mode. Connections will only be possible from clients running on this machine.
1630502750: Create a configuration file which defines a listener to allow remote access.

Nonostante io abbia un file di configurazione?
Grazie ancora

**Reply by snakuzzo** [reply]
Ciao.
Non c'entra niente il router...come ti ha detto @KC-inDomus hai un servizio sul server che è già in ascolto sulla porta 1883.
Controlla di non aver già avviato un'istanza di mosquitto

Per quanto riguarda il file di configurazione,  Ti riporto un esempio di come andrebbe scritto (in questo caso...utilizzando il `pwfile`)

```
persistence true
persistence_location /mosquitto/data
listener 1883
log_dest file /mosquitto/log/mosquitto.log
log_timestamp_format %Y-%m-%dT%H:%M:%S
# To create pwfile follow instructions at https://mosquitto.org/man/mosquitto_passwd-1.html ...
# ...and uncomment following lines to enable password authentication
allow_anonymous false
password_file /mosquitto/config/pwfile
```

...ovviamente vanno adeguati i path a seconda della tua installazione

**Reply by fdlab01** [reply]
Grazie mille snakuzzo, si avevi ragione nel fare vari tentativi disperati avevo avviato più istanze, una volte stoppato il servizioe riavviato in modalità verbose adesso ho le seguenti linee:

```
1630519587: mosquitto version 2.0.11 starting
1630519587: Using default config.
1630519587: Starting in local only mode. Connections will only be possible from clients running on this machine.
1630519587: Create a configuration file which defines a listener to allow remote access.
1630519587: For more details see https://mosquitto.org/documentation/authentication-methods/
1630519587: Opening ipv4 listen socket on port 1883.
1630519587: Opening ipv6 listen socket on port 1883.
1630519587: mosquitto version 2.0.11 running
```

Per quanto riguarda il file .conf  è il seguente:

```
pid_file /run/mosquitto/mosquitto.pid
listener 1883

persistence true
persistence_location /var/lib/mosquitto/

log_dest topic
log_type error
log_type warning
log_type notice
log_type information
connection_messages true
log_timestamp true

allow_anonymous false
password_file /etc/mosquitto/passwords

log_dest file /var/log/mosquitto/mosquitto.log
include_dir /etc/mosquitto/conf.d
```

Come guida ho utilizzata quella proposta su indomud.it al seguente link:
https://indomus.it/guide/come-installare-e-configurare-il-broker-mqtt-mosquitto-su-raspberry-pi/

Potresti darmi un aiuto in più? 
Grazie mille

**Reply by snakuzzo** [reply]
Hai provato ad adattare il "mio" file di configurazione che ho postato ieri ?

**Reply by fdlab01** [reply]
Si certo ho provato ad adattare il tuo file ma il problema è lo stesso.. 
Perdonami ma a questo punto mi ritengo un vero ignorante... forse avrei bisogno di un "tutor" che mi guidi fin dall'inizio, sicuramente mi manca qualche nozione di base!! 
Il mio file di configurazione è quello che ho postato ed è caricato correttamente come anche il log mi dice. 
I path non credo di doverli adeguare se ho seguito alla lettera la guida di indomus, ad eccezione ovviamente di username e password che sono personali il resto dovrebbe essere statico se seguo una guida con un installazione pulita su Raspbian... o sbaglio? 
Ad ogni modo se vuoi e puoi dedicarmi un po del tuo tempo sono disposto pure a ricompensarti e ripartire da un installazione pulita, a costo di imparare
Grazie

**Reply by xyzmauro** [reply]
credo tu abbia un problema coi permessi di scrittura in "pid_file /run/mosquitto/mosquitto.pid" , non ricordo come ho risolto ma verifica se è così

**Reply by snakuzzo** [reply]
@fdlab01 leggi qui
https://mosquitto.org/blog/2020/12/version-2-0-0-released/

C'è scritto tutto

**Reply by fdlab01** [reply]
@xyzmauro può essere, effettivamente non riesco nemmeno ad accedere in lettura della cartella /run/mosquitto/ ed anche da terminale utilizzando "root su" se mi sposto in quella directory e digito "ls" non mi da alcun file

@snakuzzo L'ho letto tutto questo post ma non mi è stato di aiuto, credo che il mio problema sia alla base dell'installazione ma non ne vengo a capo

**Reply by snakuzzo** [reply]
No...il tuo problema è di configurazione, non di installazione.

**Reply by snakuzzo** [reply]
Da quello che vedo, secondo me stai avviando a mano il processo.
Se stai avviando semplicemente con 

`mosquitto`

non prende il file di configurazione `/etc/mosquitto/mosquitto.conf`

e ne è la prova il fatto che ti dice di definire un listener che tu invece hai nel "tuo" file di configurazione.

Quindi dovresti usare

`sudo mosquitto -c /etc/mosquitto/mosquitto.conf`

Inoltre, se lo stai avviando con l'utente `pi`, avrai problemi perché non potrai scrivere sotto `/var/log` e `/var/lib`.

La maniera "pulita" per avviarlo invece è usando service e/o systemcl sfruttando lo "script" creato in automatico quando hai installato  `/etc/init.d/mosquitto`

Quindi, ho provato un'installazione "nuova" sul mio RPI. Questo è il file di configurazione (da adattare path e nome del `pwfile` ovviamente)
Come potrai vedere ho adeguato un po' di cose, tra cui...

- solo ipv4
- log su stdout nel caso in cui lo avviassi a mano con `sudo mosquitto...`
- timestamp formattato

```
listener 1883
socket_domain ipv4

include_dir /etc/mosquitto/conf.d
persistence true
persistence_location /var/lib/mosquitto/

log_dest topic
log_dest stdout
log_dest file /var/log/mosquitto/mosquitto.log

log_type error
log_type warning
log_type notice
log_type information
log_timestamp_format %Y-%m-%dT%H:%M:%S

connection_messages true

allow_anonymous false
password_file /etc/mosquitto/pwfile
```

e questi sono i comandi per avviarlo/fermarlo/restartarlo/controllare lo stato

```
sudo service mosquitto start
sudo service mosquitto stop
sudo service mosquitto restart
sudo service mosquitto status
```
